manager/deallocator/deallocator.go

```diff
@@ -0,0 +1,289 @@
+	resourceDeallocators[reflect.TypeOf(resourceDeallocator.event)] = resourceDeallocator
+}
+
+// A Deallocator waits for services to fully shutdown (ie no containers left)
+// and then proceeds to deallocate service-level resources (e.g. networks),
+// and finally services themselves
+// in particular, the Deallocator should be the only place where services, or
+// service-level resources, are ever deleted!
+// it’s worth noting that this new component’s role is quite different from
+// the task reaper’s: tasks are purely internal to Swarmkit, and their status
+// is entirely managed by the system itself. In contrast, the deallocator is
+// responsible for safely deleting entities that are directly controlled by the
+// user.
+type Deallocator struct {
+	store *store.MemoryStore
+
+	// for services that are shutting down, we keep track of how many
+	// tasks still exist for them
+	services map[string]*serviceWithTaskCounts
+
+	// mainly used for tests, so that we can peek
+	// into the DB state in between events
+	// the bool notifies whether any DB update was actually performed
+	eventChan chan bool
+
+	stopChan chan struct{}
+	doneChan chan struct{}
+}
+
+type serviceWithTaskCounts struct {
+	service   *api.Service
+	taskCount int
+}
+
+// New creates a new deallocator
+func New(store *store.MemoryStore) *Deallocator {
+	return &Deallocator{
+		store:    store,
+		services: make(map[string]*serviceWithTaskCounts),
+
+		stopChan: make(chan struct{}),
+		doneChan: make(chan struct{}),
+	}
+}
+
+// Run starts the deallocator, which then starts cleaning up services
+// and their resources when relevant (ie when no tasks still exist
+// for a given service)
+// This is a blocking function
+func (deallocator *Deallocator) Run(ctx context.Context) error {
+	// those are the events we'll be watching
+	events := make([]api.Event, len(resourceDeallocators)+2)
+	// we want to know when tasks are deleted by the reaper
+	events[0] = api.EventDeleteTask{}
+	// and when services get marked for deletion
+	events[1] = api.EventUpdateService{}
+	i := 2
+	for _, resourceDeallocator := range resourceDeallocators {
+		events[i] = resourceDeallocator.event
+		i++
+	}
+
+	var allServices []*api.Service
+	resourceLists := make(map[reflect.Type][]serviceLevelResource)
+
+	eventsChan, _, err := store.ViewAndWatch(deallocator.store,
+		func(readTx store.ReadTx) (err error) {
+			// look for services that are marked for deletion
+			// there's no index on the `PendingDelete` field,
+			// so we just iterate over all of them and filter manually
+			// this is okay since we only do this at leadership change
+			allServices, err = store.FindServices(readTx, store.All)
+
+			if err != nil {
+				log.G(ctx).WithError(err).Error("failed to list services in deallocator init")
+				return err
+			}
+
+			// now we also need to look at all existing service-level resources
+			// that are marked for deletion
+			for key, resourceDeallocator := range resourceDeallocators {
```

Comment:
yeah the more that i think about this... we ought to just do 

```go
allNetworks, err = store.FindNetworks(readTx, store.All)
```

and then later on if we add new resources (i.e. volumes), we'll just add more code:

```go
allVolumes, err := store.FindVolumes(readTx, store.All)
```

This does lead to duplication, but it's much more clear. If we were ever going to have more than a handful of object types that needed to be managed by this, it would be different, but we almost certainly won't.